plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java

```diff
@@ -32,11 +36,34 @@
 import org.elasticsearch.discovery.gce.GceUnicastHostsProvider;
 import org.elasticsearch.plugins.Plugin;
 
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 
 public class GceDiscoveryPlugin extends Plugin {
+    static {
+        /*
+         * GCE's http client changes access levels because its silly and we
+         * can't allow that on any old stack stack so we pull it here, up front,
+         * so we can cleanly check the permissions for it. Without this changing
+         * the permission can fail if any part of core is on the stack because
+         * our plugin permissions don't allow core to "reach through" plugins to
+         * change the permission. Because that'd be silly.
+         */
+        SecurityManager sm = System.getSecurityManager();
+        if (sm != null) {
+            sm.checkPermission(new SpecialPermission());
+        }
+        AccessController.doPrivileged(new PrivilegedAction<Void>() {
+            @Override
+            public Void run() {
+                ClassInfo.of(HttpHeaders.class, true);
+                return null;
+            }
+        });
+    }
```

Comment:
> If it involves weakhashmaps, maybe its something easy we can fix for them as well to avoid pain.

How do you mean? Like we can just keep our own strong references around?